### PR TITLE
CONSOLE-5163: Add labels field to Ingress componentRoutes

### PR DESCRIPTION
**Pull request by Leo6Leo**
## Summary 
Add a `labels` field (`map[string]string`) to `ComponentRouteSpec` in the Ingress config API (`config/v1`). This allows cluster admins to specify labels on component routes that IngressControllers use for route selection and sharding.

## Motivation 
When running multiple IngressControllers (e.g., for internal vs. external traffic), each controller uses route labels/selectors to determine which routes it manages. Currently, `componentRoutes` entries have no way to specify labels, so component routes (like the console) always land on the default IngressController. This makes it impossible to route console traffic through a specific shard without manual post-creation patching.

## Changes 
- Add `labels` field to `ComponentRouteSpec` with: - `+optional`, `+mapType=granular` for proper strategic merge patch 
- `maxProperties=8` upper bound 
- CEL validation enforcing Kubernetes label key/value format 
- Regenerated CRD manifests, OpenAPI, deepcopy, and swagger docs 
- Unit tests covering valid labels, empty map, invalid key/value rejection, maxProperties overflow, and update scenarios
- Added `+kubebuilder:validation:MaxItems=250` to the existing componentRoutes field. This is required to bound the parent list cardinality for CEL cost estimation on the nested labels map validation rules. The limit of 250 is generous — component routes are managed by OpenShift operators, and typical clusters have fewer than 10.

## Example 
```
spec:
    componentRoutes:
    - name: console-2
      namespace: openshift-console
      hostname: console.internal.corp.example.com
      labels:
        ingress: shard-console2
    - name: console-3
      namespace: openshift-console
      hostname: console.private.corp.example.com
      labels:
        ingress: shard-console3
```

### CODE REVIEW

File: config/v1/types_ingress.go [hunk 1]
```diff
@@ -245,6 +245,20 @@ type ComponentRouteSpec struct {
...
 	// +optional
 	ServingCertKeyPairSecret SecretNameReference `json:"servingCertKeyPairSecret"`
+
+	// labels defines additional labels to be applied to the route created
+	// for the component. These labels are used by the IngressController to
+	// determine which routes it should manage.
+	// Label keys and values must conform to Kubernetes label conventions:
+	// keys must be 1-63 characters (with optional prefix up to 253 characters),
+	// and values must be 0-63 characters, consisting of alphanumeric characters,
+	// '-', '_', or '.', and must start and end with an alphanumeric character.
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: The godoc needs a few additions:

1. Document the `MaxProperties=8` constraint.
2. Explain what happens when the field is omitted (required for optional fields on config APIs).
3. The comment claims "optional prefix up to 253 characters" but the CEL regex doesn't enforce the 253-char limit. Either enforce it or soften the wording.
4. Mention the operational impact: changing labels can move routes between IngressController shards.

```suggestion
	// labels defines additional labels to be applied to the route created
	// for the component. These labels are used by the IngressController to
	// determine which routes it should manage. Changing labels may cause the
	// route to be reassigned to a different IngressController.
	// When omitted, no additional labels are applied to the component route.
	// Label keys and values must conform to Kubernetes label conventions:
	// keys must be 1-63 characters (with optional DNS subdomain prefix followed
	// by a slash), and values must be 0-63 characters, consisting of alphanumeric
	// characters, '-', '_', or '.', and must start and end with an alphanumeric
	// character.
	// A maximum of 8 labels may be specified.
```

File: config/v1/types_ingress.go [hunk 2]
```diff
@@ -245,6 +245,20 @@ type ComponentRouteSpec struct {
...
+	// +optional
+	// +mapType=granular
+	// +kubebuilder:validation:MaxProperties=8
```
Comment (saschagrunert) [inline config/v1/types_ingress.go:280]: What is the rationale for limiting to 8 labels? Please document the reasoning.

File: config/v1/types_ingress.go [hunk 3]
```diff
@@ -245,6 +245,20 @@ type ComponentRouteSpec struct {
...
+	// +kubebuilder:validation:XValidation:rule="self.all(key, key.matches('^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*[/])?([A-Za-z0-9]([-A-Za-z0-9_.]{0,61}[A-Za-z0-9])?)$'))",message="label keys must be valid Kubernetes label keys"
+	// +kubebuilder:validation:XValidation:rule="self.all(key, self[key].matches('^(([A-Za-z0-9][-A-Za-z0-9_.]{0,61})?[A-Za-z0-9])?$'))",message="label values must be valid Kubernetes label values (at most 63 characters, alphanumeric, '-', '_', or '.', must start and end with alphanumeric)"
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: Consider using CEL Kubernetes library functions (`format.qualifiedName()` for keys, `format.labelValue()` for values) instead of hand-rolled regexes. They are more maintainable, cover the full Kubernetes label spec (including the 253-char prefix limit the current regex misses), and provide better error messages.

Also consider whether keys with reserved prefixes (`kubernetes.io/`, `k8s.io/`) should be rejected, since those are reserved for system use.

File: config/v1/types_ingress.go [hunk 4]
```diff
@@ -245,6 +245,20 @@ type ComponentRouteSpec struct {
...
+	Labels map[string]string `json:"labels,omitempty"`
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: Two issues here:

1. **FeatureGate required.** New fields on stable (compatibility-level-1) APIs must be behind a FeatureGate. Add a gate in `features/features.go` and mark this field with `+openshift:enable:FeatureGate=<GateName>`.

2. **`make lint` fails.** Add `+kubebuilder:validation:MinProperties=1` to prevent semantically meaningless `labels: {}`.

File: config/v1/tests/ingresses.config.openshift.io/AAA_ungated.yaml [hunk 5]
```diff
@@ -23,6 +23,129 @@ tests:
         kind: Ingress
         spec:
           domain: apps.example.com
+    - name: Should be able to create an Ingress with componentRoutes containing labels
```
Comment (saschagrunert) [inline config/v1/tests/ingresses.config.openshift.io/AAA_ungated.yaml]: Once the FeatureGate is added, these tests should move to a gate-specific file (e.g., `ComponentRouteLabels.yaml`) instead of `AAA_ungated.yaml`.

Also, please add the following test cases:
- Label key with DNS prefix (e.g., `example.com/my-label: value`), since the prefix path is the most complex part of the key regex
- Keys and values at the maximum allowed length (63 characters) to verify boundary behavior
- Update to remove labels (labels present -> no labels)
- Update to change an existing label's value

**Review by saschagrunert — COMMENTED** [review]
Thanks for working on this, the use case for label-based route sharding makes sense. See inline comments for details.

File: config/v1/types_ingress.go (hunk 2 above)
Comment (Leo6Leo) [reply config/v1/types_ingress.go:280]: The limit of 8 was chosen as a practical upper bound — in typical route sharding scenarios, IngressController selectors only need 1-2 labels, so 8 seemed generous while preventing unbounded growth. That said, there is no strong technical justification for this specific number.

Would you prefer we remove the limit entirely, or do you have a recommended ceiling for map fields like this? Happy to adjust. @saschagrunert

Comment (saschagrunert) [reply config/v1/types_ingress.go:280]: 8 is fine. Maps need a bound for CEL cost estimation and to prevent unbounded growth. For route sharding where 1-2 labels is typical, 8 is generous enough. No need to change it.

File: features/features.go [hunk 6]
```diff
@@ -638,6 +638,15 @@ var (
 								enable(inDevPreviewNoUpgrade(), inTechPreviewNoUpgrade()).
 								mustRegister()
 
+	// TODO(leoli): Replace with actual enhancement PR URL once filed. currently consulting the team for next step
+	FeatureGateIngressComponentRouteLabels = newFeatureGate("IngressComponentRouteLabels").
+						reportProblemsToJiraComponent("Management Console").
+						contactPerson("leoli").
+						productScope(ocpSpecific).
+						enhancementPR("https://github.com/openshift/enhancements/pull/XXXX").
```
Comment (saschagrunert) [inline features/features.go]: Placeholder URL. This needs a real enhancement PR before merge.

File: config/v1/types_ingress.go [hunk 7]
```diff
@@ -245,6 +245,25 @@ type ComponentRouteSpec struct {
...
+	// for Kubernetes use and may not be specified.
+	// A maximum of 8 labels may be specified.
+	// +openshift:enable:FeatureGate=IngressComponentRouteLabels
+	// +optional
+	// +mapType=granular
+	// +kubebuilder:validation:MinProperties=1
+	// +kubebuilder:validation:MaxProperties=8
+	// +kubebuilder:validation:XValidation:rule="self.all(key, !format.qualifiedName().validate(key).hasValue())",message="label keys must be valid Kubernetes qualified names"
+	// +kubebuilder:validation:XValidation:rule="self.all(key, !key.startsWith('kubernetes.io/') && !key.startsWith('k8s.io/'))",message="label keys must not use reserved prefixes kubernetes.io/ or k8s.io/"
+	// +kubebuilder:validation:XValidation:rule="self.all(key, self[key].matches('^(([A-Za-z0-9][-A-Za-z0-9_.]{0,61})?[A-Za-z0-9])?$'))",message="label values must be valid Kubernetes label values (at most 63 characters, alphanumeric, '-', '_', or '.', must start and end with alphanumeric)"
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: Use `format.labelValue()` instead of the hand-rolled regex, consistent with how you use `format.qualifiedName()` for keys:

```suggestion
	// +kubebuilder:validation:XValidation:rule="self.all(key, !format.labelValue().validate(self[key]).hasValue())",message="label values must be valid Kubernetes label values (at most 63 characters, alphanumeric, '-', '_', or '.', must start and end with alphanumeric)"
```

File: config/v1/types_ingress.go [hunk 8]
```diff
@@ -245,6 +245,25 @@ type ComponentRouteSpec struct {
...
+
+	// labels defines additional labels to be applied to the route created
+	// for the component. These labels are used by the IngressController to
+	// determine which routes it should manage. Changing labels may cause the
+	// route to be reassigned to a different IngressController.
+	// When omitted, no additional labels are applied to the component route.
+	// Label keys and values must conform to Kubernetes label conventions.
+	// Keys with the "kubernetes.io/" and "k8s.io/" prefixes are reserved
+	// for Kubernetes use and may not be specified.
+	// A maximum of 8 labels may be specified.
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: The godoc should also document `MinProperties=1` (that an empty map is not valid when the field is set). Something like:

```suggestion
	// When specified, labels must contain at least one entry, up to a maximum of 8.
```

File: payload-manifests/crds/0000_10_config-operator_01_ingresses-Default.crd.yaml [hunk 9]
```diff
@@ -7,6 +7,7 @@ metadata:
     include.release.openshift.io/ibm-cloud-managed: "true"
     include.release.openshift.io/self-managed-high-availability: "true"
     release.openshift.io/bootstrap-required: "true"
+    release.openshift.io/feature-set: Default
```
Comment (saschagrunert) [inline payload-manifests/crds/0000_10_config-operator_01_ingresses-Default.crd.yaml:10]: This looks manually edited. With the FeatureGate splitting the CRD into feature-set variants under `zz_generated.crd-manifests/`, running `make update` (which calls `hack/update-payload-crds.sh`) should produce separate payload-manifest files per feature set (like `authentications` does). Run the full `make update` and let the script handle this.

File: config/v1/tests/ingresses.config.openshift.io/IngressComponentRouteLabels.yaml [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+                label2: value2
+                label3: value3
+                label4: value4
+                label5: value5
+                label6: value6
+                label7: value7
+                label8: value8
+                label9: value9
+      expectedError: "Too many properties"
+  onUpdate:
```
Comment (saschagrunert) [inline config/v1/tests/ingresses.config.openshift.io/IngressComponentRouteLabels.yaml:268]: The tests cover basic CRUD and some negatives but the CEL validations need more coverage. Please add:

- Empty string value (`ingress: ""`) accepted (valid per K8s label spec, non-obvious edge case)
- Key name part over 63 characters rejected (tests `format.qualifiedName()` enforcement)
- Value with valid characters in invalid position (e.g., `-starts-with-dash`) rejected
- Non-reserved prefix accepted (e.g., `openshift.io/my-label`), confirms the reserved check is not overly broad

The MinProperties/MaxProperties boundary tests are basic OpenAPI and do not need separate test cases.

**Review by saschagrunert — COMMENTED** [review]
Thanks for the updates, good progress addressing the earlier feedback. A few remaining items.

File: features/features.go (hunk 6 above)
Comment (Leo6Leo) [reply features/features.go]: Just got a go ahead from our team's architect, so I will draft a enhancement soon and update the PR link whenever it is available.

**Review by saschagrunert — COMMENTED** [review]
Thanks for the updates, just a few remaining items as well as the enhancement PR blocker.

File: config/v1/types_ingress.go [hunk 11]
```diff
@@ -164,6 +164,14 @@ const (
 	Classic AWSLBType = "Classic"
 )
 
+// LabelValue is the value part of a Kubernetes label.
+// A label value must be 0-63 characters, consisting of alphanumeric characters,
+// '-', '_', or '.', and must start and end with an alphanumeric character.
+// An empty string is a valid label value.
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: Per conventions, prefer CEL over `Pattern` for human-readable error messages instead of dumping regex to users. Keep the `MaxLength=63` above for CEL cost estimation.

```suggestion
	// +kubebuilder:validation:XValidation:rule="!format.labelValue().validate(self).hasValue()",message="label values must be valid Kubernetes label values (at most 63 characters, alphanumeric, '-', '_', or '.', must start and end with alphanumeric)"
```

File: config/v1/types_ingress.go [hunk 12]
```diff
@@ -245,6 +253,22 @@ type ComponentRouteSpec struct {
...
+	// Label keys and values must conform to Kubernetes label conventions.
+	// Label values are validated at the CRD level. Label keys are validated
+	// by the Route API when the route is created or updated.
+	// When specified, labels must contain at least one entry, up to a maximum of 8.
+	// +openshift:enable:FeatureGate=IngressComponentRouteLabels
+	// +optional
+	// +mapType=granular
+	// +kubebuilder:validation:MinProperties=1
+	// +kubebuilder:validation:MaxProperties=8
+	Labels map[string]LabelValue `json:"labels,omitempty"`
```
Comment (saschagrunert) [inline config/v1/types_ingress.go:283]: There is no key validation at the CRD level. Users can submit invalid keys (e.g. `"!!!": "value"`) and only get an error asynchronously when the controller creates the Route. Add CEL rules for synchronous validation:

```go
// +kubebuilder:validation:XValidation:rule="self.all(key, !format.qualifiedName().validate(key).hasValue())",message="label keys must be valid qualified names"
// +kubebuilder:validation:XValidation:rule="self.all(key, !key.startsWith('kubernetes.io/') && !key.startsWith('k8s.io/'))",message="kubernetes.io/ and k8s.io/ prefixed label keys are reserved and may not be used"
```

File: config/v1/types_ingress.go [hunk 13]
```diff
@@ -245,6 +253,22 @@ type ComponentRouteSpec struct {
...
 	ServingCertKeyPairSecret SecretNameReference `json:"servingCertKeyPairSecret"`
+
+	// labels defines additional labels to be applied to the route created
+	// for the component. These labels are used by the IngressController to
+	// determine which routes it should manage. Changing labels may cause the
+	// route to be reassigned to a different IngressController.
+	// When omitted, no additional labels are applied to the component route.
+	// Label keys and values must conform to Kubernetes label conventions.
+	// Label values are validated at the CRD level. Label keys are validated
+	// by the Route API when the route is created or updated.
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: "Validated at the CRD level" / "validated by the Route API" are implementation details, not user-facing documentation. Document the constraints themselves instead.

```suggestion
	// Label keys must be valid qualified names per Kubernetes label conventions.
	// Keys with the "kubernetes.io/" and "k8s.io/" prefixes are reserved and may not be used.
```

File: features/features.go [hunk 14]
```diff
@@ -638,6 +638,15 @@ var (
...
+						enable(inTechPreviewNoUpgrade(), inDevPreviewNoUpgrade()).
```
Comment (saschagrunert) [inline features/features.go]: Nit: rest of the file puts `inDevPreviewNoUpgrade()` first.

```suggestion
						enable(inDevPreviewNoUpgrade(), inTechPreviewNoUpgrade()).
```

**Comment by Leo6Leo** [comment]
The enhancement PR:  https://github.com/openshift/enhancements/pull/2033

File: config/v1/types_ingress.go [hunk 15]
```diff
@@ -68,6 +68,7 @@ type IngressSpec struct {
 	// +listType=map
 	// +listMapKey=namespace
 	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=250
```
Comment (saschagrunert) [inline config/v1/types_ingress.go:72]: `MaxItems=250` is a new constraint on an existing stable field. Two things needed:

1. Document it in the `componentRoutes` godoc (e.g., "A maximum of 250 component routes may be configured."). All validation markers must be reflected in the field comment.
2. Call it out in the PR description as a necessary addition for CEL cost estimation on the nested `labels` map, so reviewers can see it was intentional rather than an unrelated change.

File: config/v1/types_ingress.go [hunk 16]
```diff
@@ -164,6 +165,13 @@ const (
 	Classic AWSLBType = "Classic"
 )
 
+// LabelValue is the value part of a Kubernetes label.
+// A label value must be 0-63 characters, consisting of alphanumeric characters,
+// '-', '_', or '.', and must start and end with an alphanumeric character.
```
Comment (saschagrunert) [inline config/v1/types_ingress.go]: "must be 0-63 characters ... and must start and end with an alphanumeric character" is contradictory for the empty string case, which is valid per the Kubernetes label spec and passes `format.labelValue()`.

```suggestion
// LabelValue is the value part of a Kubernetes label.
// A label value must be either empty or 1-63 characters, consisting of
// alphanumeric characters, '-', '_', or '.', starting and ending with
// an alphanumeric character.
```

File: features/features.go [hunk 17]
```diff
@@ -638,6 +638,14 @@ var (
 								enable(inDevPreviewNoUpgrade(), inTechPreviewNoUpgrade()).
 								mustRegister()
 
+	FeatureGateIngressComponentRouteLabels = newFeatureGate("IngressComponentRouteLabels").
+						reportProblemsToJiraComponent("Management Console").
+						contactPerson("leoli").
+						productScope(ocpSpecific).
```
Comment (saschagrunert) [inline features/features.go:672]: The enhancement PR (openshift/enhancements#2033) is still open. Per FeatureGate conventions, the enhancement needs an initial ack for DevPreview and must be merged for TechPreview.

Comment (Leo6Leo) [reply features/features.go:672]: It has received the approval from @spadgett

Comment (spadgett) [reply features/features.go:672]: @saschagrunert What level of approval do you need? Does the enhancement need to be merged, or is the ack something else? I was hoping @Miciah could look from the networking side, but the approach in the enhancement looks good to me.

Comment (saschagrunert) [reply features/features.go:672]: For feature gate promotion levels, the expected approval levels are:

- **DevPreviewNoUpgrade**: enhancement needs an initial ack (confirming direction)
- **TechPreviewNoUpgrade**: enhancement must be merged
- **Default (GA)**: enhancement must detail the GA plan and represent the implementation, plus the `verify-feature-promotion` CI check must pass (5+ tests, 14 runs, 95%+ pass rate across core platforms)

Since this PR adds a new TechPreview field, the enhancement should be merged before we can approve it here.

File: config/v1/types_ingress.go (hunk 15 above)
Comment (everettraven) [reply config/v1/types_ingress.go:72]: Changes to existing field validations also require ratcheting validation test cases and evidence that a maximum of 250 is reasonable based on telemetry data we have.

File: config/v1/types_ingress.go [hunk 18]
```diff
@@ -245,6 +255,24 @@ type ComponentRouteSpec struct {
...
+	// Keys with the "kubernetes.io/" and "k8s.io/" prefixes are reserved and may not be used.
+	// When specified, labels must contain at least one entry, up to a maximum of 8.
+	// +openshift:enable:FeatureGate=IngressComponentRouteLabels
+	// +optional
+	// +mapType=granular
+	// +kubebuilder:validation:MinProperties=1
+	// +kubebuilder:validation:MaxProperties=8
+	// +kubebuilder:validation:XValidation:rule="self.all(key, !format.qualifiedName().validate(key).hasValue())",message="label keys must be valid qualified names"
+	// +kubebuilder:validation:XValidation:rule="self.all(key, !key.startsWith('kubernetes.io/') && !key.startsWith('k8s.io/'))",message="kubernetes.io/ and k8s.io/ prefixed label keys are reserved and may not be used"
+	Labels map[string]LabelValue `json:"labels,omitempty"`
```
Comment (everettraven) [inline config/v1/types_ingress.go:283]: We strongly prefer a list of objects over the use of maps. Something like:
```yaml
labels:
  - key: mykey
     value: myvalue
```

This makes validation of keys more straightforward.

If the intention is for this to be a 1:1 mapping of metadata labels to apply to a route, using a map _may_ be acceptable if that is explicitly documented, but it is still less preferred.

More context on the preference: https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

File: config/v1/types_ingress.go [hunk 19]
```diff
@@ -245,6 +255,24 @@ type ComponentRouteSpec struct {
...
 	ServingCertKeyPairSecret SecretNameReference `json:"servingCertKeyPairSecret"`
+
+	// labels defines additional labels to be applied to the route created
+	// for the component. These labels are used by the IngressController to
+	// determine which routes it should manage. Changing labels may cause the
+	// route to be reassigned to a different IngressController.
+	// When omitted, no additional labels are applied to the component route.
+	// Label keys and values must conform to Kubernetes label conventions.
+	// Label keys must be valid qualified names per Kubernetes label conventions.
+	// Keys with the "kubernetes.io/" and "k8s.io/" prefixes are reserved and may not be used.
```
Comment (everettraven) [inline config/v1/types_ingress.go]: What about keys with `openshift.io/` as a prefix? Should those be reserved?

File: config/v1/types_ingress.go (hunk 18 above)
Comment (Leo6Leo) [reply config/v1/types_ingress.go:283]: This field carries Kubernetes labels — the same data shape as metadata.labels, matchLabels, and nodeSelector, all of which use map[string]string in the Kubernetes API. The K8s API conventions doc explicitly [exempts pure label/selector/annotation maps from the "lists over maps" rule](https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventi%20ons.md#lists-of-named-subobjects-preferred-over-maps). Key validation is handled via CEL (`format.qualifiedName()`), and `+mapType=granular` ensures correct server-side apply merge behavior.

Do you think this convincing enough to stay with the map option? @everettraven

File: config/v1/types_ingress.go (hunk 15 above)
Comment (Leo6Leo) [reply config/v1/types_ingress.go:72]: I see, I added MaxItems=250 to bound the parent list cardinality for CEL cost estimation on the nested labels map. Since this tightens a constraint on an existing stable field, I'd like to explore alternatives — do you know of an approach to bound CEL cost without constraining the parent list?

If MaxItems is the only path, I can add ratcheting validation tests. On the evidence side: only two operators write ComponentRouteStatus entries — authentication-operator (oauth-openshift) and console-operator (console, downloads) — producing 3 component routes per cluster. The spec.componentRoutes field is designed for cluster admins to customize these operator-managed routes, not to create arbitrary entries, so real-world usage is bounded by the number of participating operators. 250 gives a 50x+ margin over realistic usage. If there's a telemetry source that could further validate this, could you point me to it? @everettraven

Comment (everettraven) [reply config/v1/types_ingress.go:72]: https://telemeter-lts-dashboards.datahub.redhat.com/dashboards is probably what you are looking for when it comes to looking at telemetry of clusters that are reporting telemetry (requires VPN).

If we are fairly confident that 250 is not going to impact end-users, I'm happy with 250 + ratcheting tests.

There is not a way to bound the CEL cost without constraining the list.

File: config/v1/types_ingress.go (hunk 18 above)
Comment (everettraven) [reply config/v1/types_ingress.go:283]: If you feel strongly that the map-based approach is the best experience for end users, I won't block this for the list of named subobjects.

File: features/features.go (hunk 17 above)
Comment (jhadvig) [reply features/features.go:672]: @saschagrunert [enhancement PR ](https://github.com/openshift/enhancements/pull/2033) got merged 🤘

File: config/v1/types_ingress.go [hunk 20]
```diff
@@ -245,6 +255,24 @@ type ComponentRouteSpec struct {
...
+	// Label keys and values must conform to Kubernetes label conventions.
+	// Label keys must be valid qualified names per Kubernetes label conventions.
+	// Keys with the "kubernetes.io/", "k8s.io/", and "openshift.io/" prefixes are reserved and may not be used.
+	// When specified, labels must contain at least one entry, up to a maximum of 8.
+	// +openshift:enable:FeatureGate=IngressComponentRouteLabels
+	// +optional
+	// +mapType=granular
+	// +kubebuilder:validation:MinProperties=1
+	// +kubebuilder:validation:MaxProperties=8
+	// +kubebuilder:validation:XValidation:rule="self.all(key, !format.qualifiedName().validate(key).hasValue())",message="label keys must be valid qualified names"
```
Comment (JoelSpeed) [inline config/v1/types_ingress.go]: We should expand this description to explain what a qualified name is, similar to how we did for the labelValue

File: config/v1/types_ingress.go (hunk 18 above)
Comment (JoelSpeed) [reply config/v1/types_ingress.go:283]: I think given the nature of this field mirroring labels and feeding into labels, this is more natural as a map. I think this is an exception to the lists of subobjects case

IIRC the conventions state some exceptions as pure maps and call labels out as an example of that

**Review by JoelSpeed — COMMENTED** [review]
I haven't checked the tests but here's a couple of bits to update, will continue reviewing in the morning

File: config/v1/types_ingress.go (hunk 20 above)
Comment (JoelSpeed) [reply config/v1/types_ingress.go]: From the kube docs, we need to cover the key points from 
```
Labels are key/value pairs. Valid label keys have two segments: an optional prefix and name, separated by a slash (/). The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. The prefix is optional. If specified, the prefix must be a DNS subdomain: a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed by a slash (/).
```

So

```
label keys must be valid qualified names, consisting of a name segment and an optional prefix separated by a slash. The name segment must be at most 63 characters in length and must consist only of alphanumeric characters, dashes, underscores and dots, and must start and end with alphanumeric characters. The prefix must be at most 253 characters in length and allows the same characters as the name segment. Each dot separated segment must start and end with an alphanumeric character.
```

File: config/v1/tests/ingresses.config.openshift.io/IngressComponentRouteLabels.yaml [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+        kind: Ingress
+        spec:
+          domain: apps.example.com
+          componentRoutes:
+            - name: console
+              namespace: openshift-console
+              hostname: console.example.com
+              labels:
+                ingress: "invalid value with spaces!"
+      expectedError: "Invalid value"
```
Comment (JoelSpeed) [inline config/v1/tests/ingresses.config.openshift.io/IngressComponentRouteLabels.yaml]: This is not a sufficient error to show that your validation is working, this could be from anything. Please include a more complete error message that shows that the particular validation failed

File: config/v1/tests/ingresses.config.openshift.io/IngressComponentRouteLabels.yaml [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+        kind: Ingress
+        spec:
+          domain: apps.example.com
+          componentRoutes:
+            - name: console
+              namespace: openshift-console
+              hostname: console.example.com
+              labels:
+                kubernetes.io/name: value
+      expectedError: "prefixed label keys are reserved"
```
Comment (JoelSpeed) [inline config/v1/tests/ingresses.config.openshift.io/IngressComponentRouteLabels.yaml]: Does this error message show which prefixes are reserved? Can we expand this to a more complete message?

**Comment by saschagrunert** [comment]
I have no other comments than those which already made. :+1: Putting it into shadow review complete then.

File: config/v1/types_ingress.go (hunk 20 above)
Comment (JoelSpeed) [reply config/v1/types_ingress.go]: Can we get this in the CEL error message too please?

**Comment by JoelSpeed** [comment]
/override ci/prow/verify-crd-schema

Pre-existing failures

/lgtm
